### PR TITLE
Use runtime.GC for reducing indexing memory & replace saveMu with atomic busy loop for race control

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### Description:
this PR includes 2 feature
1. Use runtime.GC for reducing indexing memory
2. Replace save index mutex lock with atomic loop for race control

### Environment:

- Go Version: 1.15.1
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by rinx** [comment]
How about to add options for enabling/disabling `runtime.GC` calls before/after CreateIndex and SaveIndex?

**Comment by rinx** [comment]
How do you think about this comment? @kpango 
https://github.com/vdaas/vald/pull/682#issuecomment-689241495

**Comment by kpango** [comment]
sounds nice, I'll do that for now, by the way could you please test this branch?

**Comment by kpango** [comment]
@rinx do we need divide enbale/disable configuration into before and after?
I just think about define the single configuration like a `enableProactiveGC`.

File: pkg/agent/core/ngt/service/ngt.go [hunk 1]
```diff
@@ -76,8 +76,9 @@ type ngt struct {
 	dvc  *vcaches // deletion vector cache
 
 	// statuses
-	indexing atomic.Value
-	saveMu   sync.Mutex // creating or saving index
+	indexing  atomic.Value
+	saving    atomic.Value
+	lastNoice uint64 // last number of create index execution this value prevent unnecessary saveindex.
```
Comment (rinx) [inline pkg/agent/core/ngt/service/ngt.go:81]: nice! :+1:

**Comment by rinx** [comment]
> I just think about define the single configuration like a enableProactiveGC.

I've just imagined it like that. :+1:

File: pkg/agent/core/ngt/service/ngt.go [hunk 2]
```diff
@@ -644,6 +641,18 @@ func (n *ngt) SaveIndex(ctx context.Context) (err error) {
...
 
 func (n *ngt) saveIndex(ctx context.Context) (err error) {
+	noice := atomic.LoadUint64(&n.nocie)
+	if atomic.LoadUint64(&n.lastNoice) == noice {
+		return
+	}
+	atomic.SwapUint64(&n.lastNoice, noice)
+	// wait for not indexing & not saving
+	for n.IsIndexing() || n.IsSaving() {
+	}
```
Comment (rinx) [inline pkg/agent/core/ngt/service/ngt.go:651]: i think a busy loop is not a good pattern. how about to add a sleep or use channel to wait for it?

File: internal/observability/metrics/agent/core/ngt/ngt.go [hunk 3]
```diff
@@ -116,11 +133,23 @@ func (n *ngtMetrics) View() []*metrics.View {
...
+			Aggregation: metrics.LastValue(),
+		},
 		&metrics.View{
 			Name:        "ngt_is_indexing",
 			Description: n.isIndexing.Description(),
 			Measure:     &n.isIndexing,
 			Aggregation: metrics.LastValue(),
 		},
+		&metrics.View{
+			Name:        "ngt_is_indexing",
```
Comment (rinx) [inline internal/observability/metrics/agent/core/ngt/ngt.go:149]: invalid name

Comment (rinx) [reply internal/observability/metrics/agent/core/ngt/ngt.go:149]:
```suggestion
			Name:        "ngt_is_saving",
```

**Comment by kpango** [comment]
@rinx  could you please review it? you don't need to review test code it's generated by make gotest/patch

File: charts/vald/values.yaml [hunk 4]
```diff
@@ -1285,6 +1285,9 @@ agent:
     # agent.ngt.load_index_timeout_factor -- a factor of load index timeout.
     # timeout duration will be calculated by (index count to be loaded) * (factor).
     load_index_timeout_factor: 1ms
+    # @schema {"name": "agent.ngt.enable_proactive_gc", "type": "boolean"}
+    # agent.ngt.enable_proactive_gc -- proactive gc enabled
```
Comment (rinx) [inline charts/vald/values.yaml:1289]:
```suggestion
    # agent.ngt.enable_proactive_gc -- enable proactive GC call for reducing heap memory allocation
```

File: pkg/agent/core/ngt/service/ngt.go (hunk 2 above)
Comment (rinx) [reply pkg/agent/core/ngt/service/ngt.go:664]: Great! thanks for revising.

**Comment by rinx** [comment]
Others LGTM.
The CI run for Helm lint is failed because of Helm's bug. so please ignore it.

**Comment by rinx** [comment]
and please write the description of this PR.

**Pull request merged**